### DISCUSSION TITLE
Implementing Knowledge Review for Object detection

### DISCUSSION THREAD

**Discussion by ai21mtech13002** [yoshitomo-matsubara/torchdistill#287] [Q&A] [answered]
Hi Yoshitomo, 

I am now trying to implement Knowledge Review using the sample config shared by you,[https://github.com/yoshitomo-matsubara/torchdistill/blob/main/configs/sample/ilsvrc2012/single_stage/kr/resnet18_from_resnet34.yaml](url)

Can you please explain the following wrt Student4KnowlegeReview parameters:
(a) How do we set the in_channels, mid_channels and out_channels for various paths?
(b) Also what is meant by sizes: [1, 7, 14, 28, 56] and how to set those?

Thanks and Regards.

**Comment by yoshitomo-matsubara** [comment, marked as answer]
Hi @ai21mtech13002 

> (a) How do we set the in_channels, mid_channels and out_channels for various paths?

It is not my choice. I just followed [the authors' implementation](https://github.com/dvlab-research/ReviewKD/blob/master/ImageNet/models/reviewkd.py#L74-L89).

> (b) Also what is meant by sizes: [1, 7, 14, 28, 56] and how to set those?

The `sizes` are used for [upsampling](https://github.com/yoshitomo-matsubara/torchdistill/blob/8b023bb44590dd06b6dd236ee49264c907eabef9/torchdistill/models/wrapper.py#L441) by [each of the ABF modules ](https://github.com/yoshitomo-matsubara/torchdistill/blob/8b023bb44590dd06b6dd236ee49264c907eabef9/torchdistill/models/wrapper.py#L483-L484)

**Reply by ai21mtech13002** [reply]
Thanks Yoshitomo,

I am using Faster RCNN (Resnet 50 backbone as Teacher and Resnet 18 backbone as Student) to implement Knowledge Review. I am specifying the Layer 1 to 4 of each Teacher and Student in the module paths in the yaml file.

I am a bit confused as to how the shapes of the Teacher Feature Maps and Student Feature Maps will be matched for calculation of loss?

As a Student Feature Map at a level is used with every previous Teacher Student Map, how will this be taken care of? Is there a requirement of specifying a special module as was done in the case of Attention Transfer?

Thanks and Regards.

**Reply by yoshitomo-matsubara** [reply]
Hi @ai21mtech13002 

You'd need to have teacher's out_channels match those of students, adjusting in_channels for teacher/student models accordingly (potentially sizes as well)

For design choices of this specific method, you'd be better to ask the authors such questions at their repository (linked above)

**Reply by ai21mtech13002** [reply]
Hi Yoshitomo,

I have been able to match the shapes of outputs of ABFs and that of the Teacher layer outputs for the Knowledge Review process to start running.

1. The problem that I am facing now is that the training loop is getting broken as loss is becoming NaN.

I am enclosing the log for your reference.
[log.txt](https://github.com/yoshitomo-matsubara/torchdistill/files/10840981/log.txt)

2. Also can you clarify the role of kernel_sizes passed as a parameter for the HCL loss?


Thanks and Regards.

**Reply by ai21mtech13002** [reply]
Hi Yoshitomo,

I was able to resolve the problem of loss becoming NaN. I had set the learning rate too high. Reducing the same resolved it.

Can you however clarify the second point i.e. kernel_sizes as a parameter for the HCL loss?

Thanks and Regards.

**Reply by yoshitomo-matsubara** [reply]
Hi @ai21mtech13002 

The `kernel_sizes` are from https://github.com/dvlab-research/ReviewKD/blob/master/ImageNet/models/reviewkd.py#L98
As mentioned above, please ask the authors of the paper for more details
https://github.com/dvlab-research/ReviewKD
